xen/arch/arm/coproc/coproc.c

```diff
@@ -0,0 +1,381 @@
+           opt_dom0_coprocs, d->domain_id);
+
+    /*
+     * For the moment, we'll create vcoproc for each registered coproc
+     * which is described in the list of coprocs for domain 0 in bootargs.
+     */
+    for ( curr = opt_dom0_coprocs; curr; curr = next )
+    {
+        struct dt_device_node *node = NULL;
+        char *buf;
+        bool_t is_alias = false;
+
+        next = strchr(curr, ',');
+        if ( next )
+        {
+            len = next - curr;
+            next++;
+        }
+        else
+            len = strlen(curr);
+
+        if ( *curr != '/' )
+            is_alias = true;
+
+        buf = xmalloc_array(char, len + 1);
+        if ( !buf )
+        {
+            ret = -ENOMEM;
+            break;
+        }
+
+        strlcpy(buf, curr, len + 1);
+        if ( is_alias )
+            node = dt_find_node_by_alias(buf);
+        else
+            node = dt_find_node_by_path(buf);
+        if ( !node )
+        {
+            printk("Unable to find node by %s \"%s\"\n",
+                   is_alias ? "alias" : "path", buf);
+            ret = -EINVAL;
+        }
+        xfree(buf);
+        if ( ret )
+            break;
+
+        curr = dt_node_full_name(node);
+
+        ret = coproc_find_and_attach_to_domain(d, curr);
+        if (ret)
+        {
+            printk("Failed to attach coproc \"%s\" to dom%u (%d)\n",
+                   curr, d->domain_id, ret);
+            break;
+        }
+    }
+
+    return ret;
+}
+
+int vcoproc_domain_init(struct domain *d)
+{
+    struct vcoproc *vcoproc_d = &d->arch.vcoproc;
+    int ret = 0;
+
+    vcoproc_d->num_instances = 0;
+    INIT_LIST_HEAD(&vcoproc_d->instances);
+    spin_lock_init(&vcoproc_d->lock);
+
+    /*
+     * We haven't known yet if the guest domain are going to use coprocs.
+     * So, just return okay for the moment. It won't be an issue later if
+     * guest domain doesn't request any.
+     * But we definitely know when domain 0 is being created.
+     */
+    if ( !num_coprocs )
+    {
+        if ( d->domain_id == 0 && strcmp(opt_dom0_coprocs, "") )
+        {
+            printk("There is no registered coproc for creating vcoproc\n");
+            return -ENODEV;
+        }
+
+        return 0;
+    }
+
+    /* We already have the list of coprocs for domain 0 only. */
+    if ( d->domain_id == 0 )
+        ret = vcoproc_dom0_init(d);
+
+    return ret;
+}
+
+void vcoproc_domain_free(struct domain *d)
+{
+    coproc_release_vcoprocs(d);
+}
+
+int coproc_release_vcoprocs(struct domain *d)
+{
+    struct vcoproc *vcoproc_d = &d->arch.vcoproc;
+    struct vcoproc_instance *vcoproc, *temp;
+    int ret = 0;
+
+    spin_lock(&vcoproc_d->lock);
+
+    if ( list_empty(&vcoproc_d->instances) )
+        goto out;
+
+    list_for_each_entry_safe( vcoproc, temp, &vcoproc_d->instances, instance_elem )
+    {
+        spin_unlock(&vcoproc_d->lock);
+
+        ret = coproc_detach_from_domain(d, vcoproc);
```

Comment:
is it required that coproc_detach_from_domain holds the lock?
can it be w/o lock and users will lock themselves?